### PR TITLE
search: facilitate simple searches if globbing is active

### PR DESCRIPTION
**Pull request by stefanhengl**
Relates to https://github.com/sourcegraph/sourcegraph/issues/12476

With this PR we slightly modify the interpretation of filter values if globbing is active.

Users can now enter queries like

```
repo:sourcegraph file:search.go 
```

which will be interpreted as

```
repo:**sourcegraph** file:**search.go** 
```

in the backend. 

Filter values containing glob symbols or path separators will be interpreted as glob-pattern without further modification. Results that match the file: filter value exactly (e.g. files named search.go in any directory) will be listed first. Multiple exact matches are sorted by length of the file path, and then alphabetically.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/search_results.go [hunk 1]
```diff
@@ -1889,14 +1899,75 @@ func compareSearchResults(a, b SearchResultResolver) bool {
...
+		})
+	return m
+}
+
+// sortResultsAndOr is like sortResults, but fine tunes the comparison function by taking into account:
+// - the original query
+// - whether globbing is active
+//
+// If globbing is active and the original query contains exact file patterns (no glob syntax), results with
+// exact matches will appear first within their repository.
```
Comment (rvantonder) [inline cmd/frontend/graphqlbackend/search_results.go:1962]: "fine tunes" is a bit vague. Maybe a description like 

> sortResultsAndOr partitions a result set to matches that correspond to an exact file path match (ordered first in the resulting set) and matches that correspond to substring matches (ordered after exact matches).

I don't think that I captured what the function does accurately in my suggestion, I just think the words like "partition" and "order" probably conveys what's going on here a bit better.

Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search_results.go:1962]: Also thinking we can remove `AndOr` parts in these function names, it feels kinda meaningless (in theory it's not specific to `AndOr` queries right, we could have done it with existing queries? Maybe something like `sortResultsExactMatchesOnTop` but less verbose, can't think of something better right now.

File: cmd/frontend/graphqlbackend/search_results.go [hunk 2]
```diff
@@ -1889,14 +1899,75 @@ func compareSearchResults(a, b SearchResultResolver) bool {
...
+func (r *searchResolver) sortResultsAndOr(ctx context.Context, rr []SearchResultResolver) {
+	var exactPatterns map[string]struct{}
+	if settings, err := decodedViewerFinalSettings(ctx); err != nil || getBoolPtr(settings.SearchGlobbing, false) {
+		exactPatterns = r.getExactFilePatterns()
+	}
+	sort.Slice(rr, func(i, j int) bool { return compareSearchResultsAndOr(rr[i], rr[j], exactPatterns) })
+}
```
Comment (rvantonder) [inline cmd/frontend/graphqlbackend/search_results.go:1969]: I think it's fine to guard this behind the globbing option. If we decide not to use globbing, we can apply the same sorting idea to our current regular expressions right? Just confirming that if that happened, it would be straightforward to activate this functionality by removing the `SearchGlobbing` flag here (and assuming we implement a similar `getExactFilePatterns` for a mode where regex is enabled).

So basically, it's worth separating the parts where we rely on globbing to activate the sorting, versus the part that actually sorts/orders the results (`compareSearchResultsAndOr`) It seems like the code is laid out for that already.

File: cmd/frontend/graphqlbackend/search_results.go [hunk 3]
```diff
@@ -1870,16 +1875,21 @@ func isContextError(ctx context.Context, err error) bool {
...
 
-	// SearchResultURIs returns the repo name and file uri respectiveley
-	searchResultURIs() (string, string)
 	resultCount() int32
 }
 
+type searchResultURIGetter interface {
+	// SearchResultURIs returns the repo name and file uri respectively
+	searchResultURIs() (string, string)
+}
```
Comment (rvantonder) [inline cmd/frontend/graphqlbackend/search_results.go:1891]: I'm probably missing something, why is it needed to wrap `searchResultURIs` in this interface?

File: internal/search/query/searchquery.go [hunk 4]
```diff
@@ -27,6 +27,8 @@ const (
 	FieldContent            = "content"
 	FieldVisibility         = "visibility"
 
+	FieldFileOffset = len(FieldFile) + 1
```
Comment (rvantonder) [inline internal/search/query/searchquery.go:30]: Nit: we want to depend as little as possible on `searchquery.go` or add new additions here. I think it's quite fine to just inline `len(Fieldfile) + 1` in the single place this is used.

File: internal/search/query/transformer.go [hunk 5]
```diff
@@ -210,6 +216,19 @@ func reporevToRegex(value string) (string, error) {
 	return value, nil
 }
 
+var globSymbols = lazyregexp.New(`[][*?/]`)
+
+func ContainsNoGlobSymbols(value string) bool {
```
Comment (rvantonder) [inline internal/search/query/transformer.go:221]: Nit: is "syntax" better than "symbols", potentially? Or is "symbols" preferred in the context of the globbing language?

File: cmd/frontend/graphqlbackend/search_results.go [hunk 6]
```diff
@@ -1889,14 +1899,75 @@ func compareSearchResults(a, b SearchResultResolver) bool {
...
+// length of their file path and then alphabetically.
+func compareSearchResultsAndOr(a, b searchResultURIGetter, exactFilePatterns map[string]struct{}) bool {
+	arepo, afile := a.searchResultURIs()
+	brepo, bfile := b.searchResultURIs()
+
+	if arepo == brepo {
+		if exactFilePatterns == nil || len(exactFilePatterns) == 0 {
+			return afile < bfile
+		}
+		_, aMatch := exactFilePatterns[filepath.Base(afile)]
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_results.go:1920]: I think you should use `path` here, not `filepath`. As a benefit `path.Base` is more performant due to assumptions it can make. We are abstracted away from the FS (I hope :P).

File: cmd/frontend/graphqlbackend/search_results.go [hunk 7]
```diff
@@ -1841,7 +1842,11 @@ func (r *searchResolver) doResults(ctx context.Context, forceOnlyResultType stri
 		multiErr = nil
 	}
 
-	sortResults(results)
+	if _, isAndOr := r.query.(*query.AndOrQuery); isAndOr {
+		r.sortResultsAndOr(ctx, results)
+	} else {
+		sortResults(results)
+	}
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_results.go:1849]: I'm not sure why you need a different sort function for and/or. Can't we just modify sortResults?

File: cmd/frontend/graphqlbackend/search_results.go (hunk 3 above)
Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search_results.go:1891]: I extracted it from `SearchResultResolver` because the comparison function we use in sorting really only needs `searchResultURIs()`. The small interface makes testing and reuse easier.

File: cmd/frontend/graphqlbackend/search_results.go [hunk 8]
```diff
@@ -1889,14 +1899,75 @@ func compareSearchResults(a, b SearchResultResolver) bool {
...
 
+// compareSearchResultsAndOr is like compareSearchResults, but overrides sorting in alphabetical order if
+// one of the filenames is contained in exactFilePatterns, in which case exact matches are sorted by
+// length of their file path and then alphabetically.
+func compareSearchResultsAndOr(a, b searchResultURIGetter, exactFilePatterns map[string]struct{}) bool {
+	arepo, afile := a.searchResultURIs()
+	brepo, bfile := b.searchResultURIs()
+
+	if arepo == brepo {
+		if exactFilePatterns == nil || len(exactFilePatterns) == 0 {
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_results.go:1917]: you don't need to check for nil here, you can do len on nil maps.

File: internal/search/query/transformer.go (hunk 5 above)
Comment (stefanhengl) [reply internal/search/query/transformer.go:221]: let's go with "syntax"

File: cmd/frontend/graphqlbackend/search_results.go [hunk 9]
```diff
@@ -1889,14 +1899,75 @@ func compareSearchResults(a, b SearchResultResolver) bool {
...
+				}
+				if len(bfile) < len(afile) {
+					return false
+				}
+				return afile < bfile
+			}
+			if aMatch {
+				return true
+			}
+			return false
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_results.go:1935]:
```suggestion
            // prefer exact matches
			return aMatch
```

File: cmd/frontend/graphqlbackend/search_results.go [hunk 10]
```diff
@@ -1889,14 +1899,75 @@ func compareSearchResults(a, b SearchResultResolver) bool {
...
+		_, aMatch := exactFilePatterns[filepath.Base(afile)]
+		_, bMatch := exactFilePatterns[filepath.Base(bfile)]
+		if aMatch || bMatch {
+			if aMatch && bMatch {
+				if len(afile) < len(bfile) {
+					return true
+				}
+				if len(bfile) < len(afile) {
+					return false
+				}
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_results.go:1929]: nit:
```suggestion
                // Prefer shorter file names (ie root files come first)
				if len(afile) != len(bfile) {
					return len(afile) < len(bfile)
				}
```

File: cmd/frontend/graphqlbackend/search_results.go (hunk 2 above)
Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search_results.go:1969]: Yes, we could use the same functionality without globbing. The guard just protects `getExactFilePatterns` which only works with glob patterns (for now). In the future, we could move the guard inside `getExactFilePatterns` and handle the matching of glob syntax or regex syntax transparently from the perspective of the sort function.

File: cmd/frontend/graphqlbackend/search_results.go (hunk 1 above)
Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search_results.go:1962]: What about `upsortExactMatches` and `searchResultIsLess`?

Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search_results.go:1962]: After reading Keagan's comments I went for updating the original `sortResults` instead of having a dedicated sort function for `AndOr`.

File: cmd/frontend/graphqlbackend/search_results.go (hunk 7 above)
Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search_results.go:1849]: I agree, let's just update `sortResults`!
